packages/x-data-grid/src/hooks/features/density/useGridDensity.tsx

```diff
@@ -42,14 +42,18 @@ export const useGridDensity = (
           return state;
         }
 
+        if (props.onDensityChange) {
+          props.onDensityChange(newDensity);
+        }
+
         return {
           ...state,
           density: newDensityState,
         };
       });
       apiRef.current.forceUpdate();
     },
-    [logger, apiRef],
+    [logger, apiRef, props],
```

Comment:
Could it be `useEventCallback` to avoid risking errors like putting `props` here?